weaveworks/pipeline-controller
# Add prototypical promotion server

This PR adds an HTTP server that can be used as a [Flux notification Provider](https://fluxcd.io/flux/components/notification/provider/#generic-webhook) for Pipeline workloads. Each environment of each Pipeline has its own webhook URL for triggering a promotion:

http://_promotion-server-domain_/promotion/_namespace_/_name_/_environment_

The tuple (_namespace_, _name_) of the URL path identifies the Pipeline resource on the management cluster while _environment_ denotes the environment within the Pipeline that triggered the promotion. The webhook server then finds the next environment in the Pipeline and passes it to the configured promotion strategy. A nop promotion strategy is included in the PR for completeness.

Manual testing:

1. `kind create cluster`
1. `make install`
1. `kubectl apply -f config/testdata/crds/gitops_cluster.yaml`
1. `kubectl apply -f config/testdata/pipeline.yaml`
1. `go run main.go`
1.  `curl -i localhost:8082/promotion/default/podinfo/dev -d '{
    "involvedObject": {
        "apiVersion": "helm.toolkit.fluxcd.io/v2beta1",
        "kind": "HelmRelease",
        "name": "podinfo",
        "namespace": "podinfo"
    },
    "metadata": {
        "revision": "4.0.0"
    }
}'`

After the last step the following line should be logged by the server:

```
1.6629751580661228e+09  INFO    promoting app   {"promotion": {"appNS":"default","appName":"podinfo","environment":{"name":"prod","targets":[{"namespace":"podinfo","clusterRef":{"kind":"GitopsCluster","name":"prod"}}],"gitSpec":{"url":"https://git@github.com/org/prod-repo"}},"version":"4.0.0"}}
```

This PR is part 1 of a 2-part change:

1. add promotion server (this PR)
2. create notification-controller objects in each target (follow-up PR)

refs weaveworks/weave-gitops-enterprise#1487
